aggregator/src/test/java/org/jenkinsci/plugins/workflow/BuildWatcher.java

```diff
@@ -0,0 +1,153 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow;
+
+import hudson.Extension;
+import hudson.console.AnnotatedLargeText;
+import hudson.console.LineTransformationOutputStream;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.junit.rules.ExternalResource;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+/**
+ * Echoes build output to standard error as it arrives.
+ * Usage: <pre>{@code @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();}</pre>
+ * Should work in combination with {@link JenkinsRule} or {@link RestartableJenkinsRule}.
+ */
+public final class BuildWatcher extends ExternalResource {
+
+    private static boolean active;
+    private static final Map<File,RunningBuild> builds = new ConcurrentHashMap<File,RunningBuild>();
+
+    private Thread thread;
+
+    @Override protected void before() throws Throwable {
+        active = true;
+        thread = new Thread("watching builds") {
+            @Override public void run() {
+                try {
+                    while (active) {
+                        for (RunningBuild build : builds.values()) {
+                            build.copy();
+                        }
+                        Thread.sleep(50);
+                    }
+                } catch (InterruptedException x) {
+                    // stopped
+                }
+                // last chance
+                for (RunningBuild build : builds.values()) {
+                    build.copy();
+                }
+            }
+        };
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @Override protected void after() {
+        active = false;
+        thread.interrupt();
+    }
+
+    @Extension public static final class Listener extends RunListener<Run<?,?>> {
+
+        @Override public void onStarted(Run<?,?> r, TaskListener listener) {
+            RunningBuild build = new RunningBuild(r);
+            RunningBuild orig = builds.put(r.getLogFile(), build);
```

Comment:
I did not want to use the `Run` instance because this needs to work across Jenkins restarts. Nor did I want to use the `externalizableId` because I wanted this to work as a `ClassRule` across different test cases within the same suite, which often reuse the same project name and of course build number (`p #1`) for example. Whereas the `logFile` includes the randomized `$JENKINS_HOME` prefix and so should be both consistent across restart and unique across tests.